<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"></ul></div>

В аптеки города N поступил экспресс-тест на COVID-19, у которого вероятность ошибки (как позитивной, так и негативной) равна 0.05. По официальной статистике, прямо сейчас в городе N всего 1% жителей болен ковидом. Один из жителей, не имея симптомов, просто на всякий случай решил сдать этот тест, и внезапно тест показал положительный результат. Оцените вероятность того, что этот житель действительно болен ковидом.

In [1]:
import pandas as pd
import numpy as np
from tqdm.notebook import tqdm

In [39]:
# вероятность ошибки covid-теста равна 0.05
err_rate = 0.05

# 1% жителей болен ковидом
covid_probability = 0.01

In [40]:
# моделирует covid-тест
def test(covid):
    if np.random.binomial(1, 1-err_rate):
        return covid
    return abs(covid - 1)

In [41]:
# проверяем covid-тест на 10 тысячах итераций
np.mean([test(1) for _ in range(10000)])

0.949

In [67]:
# моделируем город с населением в миллион
result = []
for _ in tqdm(range(1000000)):
    # моделируем пациента, который болен с вероятностью 1%
    patient = np.random.binomial(1, covid_probability)
    # и результат его теста
    test_result = test(patient)
    result.append((patient, test_result))

  0%|          | 0/1000000 [00:00<?, ?it/s]

In [70]:
# из полученных данных создаем датафрейм
city = pd.DataFrame(result, columns=['patient', 'test_result'])
city

patient  test_result
0             0            0
1             0            1
2             0            0
3             0            1
4             0            0
...         ...          ...
999995        0            0
999996        0            0
999997        0            0
999998        0            0
999999        0            0

[1000000 rows x 2 columns]

In [76]:
# группируем по результату тестов
city.groupby('test_result')['patient'].mean()

test_result
0    0.000505
1    0.159999
Name: patient, dtype: float64

Ответ: Вероятность того, что пациент, не имеющий симптомов и сделавший тест с погрешностью 5%, действительно болен — около 16%.